lib/spack/spack/test/url_fetch.py

```
@@ -358,3 +358,84 @@ def _which(*args, **kwargs):
             with Stage(fetcher, path=testpath) as stage:
                 out = stage.fetch()
             assert err_fmt.format('curl') in out
+
+
+def test_url_missing_curl(tmpdir, monkeypatch):
+    """Check if URL's curl is missing."""
+    err_fmt = 'No such command {0}'
+
+    def _which(*args, **kwargs):
+        err_msg = err_fmt.format(args[0])
+        raise spack.util.executable.CommandNotFoundError(err_msg)
+
+    # Patching the 'which' symbol imported by url works since it is
+    # too late in import processing to patch the defining
+    # (spack.util.executable) module's symbol.
+    monkeypatch.setattr(spack.url, 'which', _which)
+
+    with spack.config.override('config:url_fetch_method', 'curl'):
+        exists = spack.url.url_exists('https://github.com/')
```

I have a couple of concerns:

* IMO `url_exists` should throw an exception if it has no tools available to verify the existence of the URL
* (this might be resolved as part of resolving the first item) if possible this test should avoid depending on "real" URLs since they could cause intermittent test failures. I think the simplest (although not perhaps the best) way to address that would be to use a `file://` URL

---

Thanks.  Let me revisit this then.

---

Note `spack.util.web.url_exists` does not re-raise exceptions from `read_from_url()` so the affects from a combined method may affect existing assumptions regarding its use.